File: Python/650-2 Keys-Keyboard.py
# https://leetcode.com/problems/2-keys-keyboard
from typing import *
import math

class Solution:
    """
    Idea: (originally done in c++)
    A number n can be reached in as many steps as the sum of its
        prime factorization and a prime number n' requires a minimum of
        n' steps. We can break n down into its prime factorization and 
        sum the values together to get a minimum number of steps to reach n.
    
    Firstly, check if n is 1, if it is, there are no steps needed. Otherwise
        we can recursively count the steps by breaking down n into a prime
        factor tree. To do this, for ever call to minSteps(n) we check if 
        n is divisible by any number in the range [2,sqrt(n)]. If it is, then
        then we recursively call minSteps on the divisor of n and the resulting
        quotient. The tree traversal will end with the base case of a number
        reaching all of its prime factors.
    """
    def minSteps(self, n: int) -> int:
        if n == 1: return 0
        for i in range(2,int(math.sqrt(n))+1):
            if n % i == 0:
                return self.minSteps(i) + self.minSteps(n//i)
        return n

File: Cpp/650-2 Keys-Keyboard.cpp
// https://leetcode.com/problems/2-keys-keyboard/

#include <bits/stdc++.h>
using namespace std;
class Solution {
    /*
    Idea:
    The way to get exactly n characters in the minimum steps involves finding out
        the prime factorization of n. Every prime number n' characters takes a minimum of n' 
        steps to achieve (copy + (n'-1)*paste). Since every non-prime number can be broken 
        down into its factorized primes, then the minimum steps to make up these primes 
        is the minimum steps to make the original number n.
        
    
    Create a helper function (count_and_factor(x)) that recursively counts up the
        number of steps needed to get x characters. This is done by iterating through every
        number in the range [2,sqrt(x)] and seeing if x is evenly divisible by any number
        in the range. If it is, then we can recursively call the function on the divisor
        and quotient. However, if x is not evenly divisible by any number in the set range,
        then that means that x must be prime and as a result, returns x steps. 
        
    Note: if n is 1, then there are zero steps to take since there is already a character.
    */
public:
    int minSteps(int n) {
        if (n == 1) return 0;
        
        return count_and_factor(n);
        
    }
    
    int count_and_factor(int x) {
        for (int i = 2; i < (int)sqrt(x)+1; i++) {
            if (x % i == 0) {
                // recursive call
                return (count_and_factor(i) + count_and_factor(x/i));               
            } 
        }
        // prime number, returns x
        return x;
    }
};